lesson-04-memory-management/mm/xxx.c

```diff
@@ -5,25 +5,136 @@
 #include <linux/pci.h>
 #include <linux/version.h>
 #include <linux/init.h>
+#include <linux/slab.h>
 
-#define LEN_MSG 160
-static char buf_msg[LEN_MSG + 1] = "Hello from module!\n";
+#define MEM_TYPE_KMALLOC	0
+#define MEM_TYPE_KCACHE 	1
+#define MEM_TYPE_KVMALLOC	2
+
+#define BUFFER_SIZE			160
+
+static size_t	buf_size = 0;
+static char 	*pbuf_msg = 0;
+static int  	mem_type = MEM_TYPE_KMALLOC;
+struct KCACHE_STRUCT {
+	char buf[BUFFER_SIZE];
+};
+
+static struct kmem_cache* pmem_cache = NULL;
+
+const char* get_mem_type(int type)
+{
+	switch (type) {
+		case MEM_TYPE_KMALLOC:
+			return "MEM_TYPE_KMALLOC";
+		case MEM_TYPE_KCACHE:
+			return "MEM_TYPE_KCACHE";
+		case MEM_TYPE_KVMALLOC:
+			return "MEM_TYPE_KVMALLOC";
+	}
+	return "UNKNOWN";
+}
+
+static void* mem_alloc(int type, size_t size)
+{
+	void* buf = NULL;
+	switch (type) {
+		case MEM_TYPE_KMALLOC:
+			buf = kmalloc(size, GFP_KERNEL);
+			break;
+		case MEM_TYPE_KCACHE:
+			buf = kmem_cache_alloc( pmem_cache, GFP_KERNEL );
+			break;
+		case MEM_TYPE_KVMALLOC:
+			buf = vmalloc(size);
+			break;
+	}
+
+	if (buf) buf_size = size;
+
+	pr_info("[%s:%s] {%s} %zu bytes --> %p\n", THIS_MODULE->name, __FUNCTION__, get_mem_type(type), size, buf);
+	return buf;
+}
+
+static void mem_free(int type, void* buf)
+{
+	if (!buf) {
+		pr_warning("[%s:%s] Wrong parameter %p\n", THIS_MODULE->name, __FUNCTION__, buf);
+		return;
+	}
+	switch (type) {
+		case MEM_TYPE_KMALLOC:
+			kfree(buf);
+			break;
+		case MEM_TYPE_KCACHE:
+			kmem_cache_free( pmem_cache, buf );
+			buf = kmem_cache_alloc( pmem_cache, GFP_KERNEL );
+			break;
+		case MEM_TYPE_KVMALLOC:
+			vfree(buf);
+			break;
+	}
+	pr_info("[%s:%s] {%s} bytes --> %p\n", THIS_MODULE->name, __FUNCTION__, get_mem_type(type), buf);
+}
+
+void init_kcache(void) 
+{
+	if (mem_type != MEM_TYPE_KCACHE) return;
```

Comment:
IMO, it's more logically to use opposite code structure:
```
if (mem_type == MEM_TYPE_KCACHE) {
        /* init cache */
}
```
And in this particular case - I'd move `mem_type` verification to the `x_init()`
and made call of `init_kcache()` conditional.
Otherwise `init_kcache()` should be renamed to reflect, it inits any kind of buffer (even if for other types nothing is done).